File: services/project-service/src/main/java/com/freelance/projectservice/service/ProjectService.java
package com.freelance.projectservice.service;

import com.freelance.projectservice.client.WalletClient;
import com.freelance.projectservice.dto.CreateProjectRequest;
import com.freelance.projectservice.dto.ProjectDTO;
import com.freelance.projectservice.dto.ProjectStatsDTO;
import com.freelance.projectservice.dto.UpdateProjectRequest;
import com.freelance.projectservice.exception.InsufficientTokensException;
import com.freelance.projectservice.exception.ResourceNotFoundException;
import com.freelance.projectservice.model.Project;
import com.freelance.projectservice.model.ProjectStatus;
import com.freelance.projectservice.repository.ProjectRepository;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Project Service Layer
 * 
 * Contains business logic for project operations.
 */
@Service
@Transactional
@Slf4j
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final WalletClient walletClient;

    @Value("${app.tokens-per-project:5}")
    private int tokensPerProject;

    public ProjectService(ProjectRepository projectRepository, WalletClient walletClient) {
        this.projectRepository = projectRepository;
        this.walletClient = walletClient;
    }
    
    /**
     * Create a new project. Deducts tokens from owner wallet if app.tokens-per-project > 0.
     */
    public ProjectDTO createProject(CreateProjectRequest request) {
        if (tokensPerProject > 0 && request.getOwnerId() != null) {
            boolean deducted = walletClient.deductTokens(request.getOwnerId(), tokensPerProject, "Project creation");
            if (!deducted) {
                throw new InsufficientTokensException(
                    "Insufficient token balance. Creating a project costs " + tokensPerProject + " tokens. Please buy tokens first.",
                    tokensPerProject
                );
            }
        }
        Project project = new Project();
        project.setTitle(request.getTitle());
        project.setDescription(request.getDescription());
        project.setStatus(request.getStatus() != null ? request.getStatus() : ProjectStatus.DRAFT);
        project.setOwnerId(request.getOwnerId());
        project.setBudget(request.getBudget());
        project.setDeadline(request.getDeadline());
        project.setCategory(request.getCategory());
        project.setImageUrl(request.getImageUrl());
        project.setTags(request.getTags());
        
        Project savedProject = projectRepository.save(project);
        log.info("Project created - ID: {}, Title: {}, Owner: {}, Category: {}", 
                savedProject.getId(), savedProject.getTitle(), savedProject.getOwnerId(), savedProject.getCategory());
        
        return convertToDTO(savedProject);
    }
    
    /**
     * Get project by ID
     */
    @Transactional(readOnly = true)
    public ProjectDTO getProjectById(Long id) {
        Project project = projectRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Project not found with id: " + id));
        return convertToDTO(project);
    }
    
    /**
     * Get all projects
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getAllProjects() {
        List<Project> projects = projectRepository.findAll();
        return projects.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Get projects by owner ID
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getProjectsByOwnerId(Long ownerId) {
        List<Project> projects = projectRepository.findByOwnerId(ownerId);
        return projects.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Get projects by status
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getProjectsByStatus(ProjectStatus status) {
        List<Project> projects = projectRepository.findByStatus(status);
        return projects.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Get projects by category
     */
    @Transactional(readOnly = true)
    public List<ProjectDTO> getProjectsByCategory(String category) {
        List<Project> projects = projectRepository.findByCategory(category);
        return projects.stream()
            .map(this::convertToDTO)
            .collect(Collectors.toList());
    }
    
    /**
     * Update project
     */
    public ProjectDTO updateProject(Long id, UpdateProjectRequest request) {
        Project project = projectRepository.findById(id)
            .orElseThrow(() -> new ResourceNotFoundException("Project not found with id: " + id));
        
        if (request.getTitle() != null && !request.getTitle().isBlank()) {
            project.setTitle(request.getTitle());
        }
        if (request.getDescription() != null) {
            project.setDescription(request.getDescription());
        }
        if (request.getStatus() != null) {
            project.setStatus(request.getStatus());
        }
        if (request.getBudget() != null) {
            project.setBudget(request.getBudget());
        }
        if (request.getDeadline() != null) {
            project.setDeadline(request.getDeadline());
        }
        if (request.getCategory() != null) {
            project.setCategory(request.getCategory());
        }
        if (request.getImageUrl() != null) {
            project.setImageUrl(request.getImageUrl());
        }
        if (request.getTags() != null) {
            project.setTags(request.getTags());
        }
        
        Project updatedProject = projectRepository.save(project);
        log.info("Project updated - ID: {}, Title: {}", updatedProject.getId(), updatedProject.getTitle());
        
        return convertToDTO(updatedProject);
    }
    
    /**
     * Delete project
     */
    public void deleteProject(Long id) {
        if (!projectRepository.existsById(id)) {
            throw new ResourceNotFoundException("Project not found with id: " + id);
        }
        projectRepository.deleteById(id);
        log.info("Project deleted - ID: {}", id);
    }
    
    /**
     * Get project statistics for backoffice dashboard.
     */
    @Transactional(readOnly = true)
    public ProjectStatsDTO getStatistics() {
        long total = projectRepository.count();
        Map<String, Long> byStatus = new HashMap<>();
        for (ProjectStatus status : ProjectStatus.values()) {
            byStatus.put(status.name(), projectRepository.countByStatus(status));
        }
        Map<String, Long> byCategory = new HashMap<>();
        List<Object[]> categoryRows = projectRepository.countByCategory();
        for (Object[] row : categoryRows) {
            String category = (String) row[0];
            Long count = (Long) row[1];
            byCategory.put(category, count);
        }
        BigDecimal totalBudget = projectRepository.sumBudget() != null ? projectRepository.sumBudget() : BigDecimal.ZERO;
        long completedCount = projectRepository.countByStatus(ProjectStatus.COMPLETED);
        double completionRate = total > 0 ? (completedCount * 100.0) / total : 0;
        long withBudget = projectRepository.countByBudgetIsNotNull();
        double averageBudget = withBudget > 0 ? totalBudget.doubleValue() / withBudget : 0;
        
        return ProjectStatsDTO.builder()
            .total(total)
            .byStatus(byStatus)
            .byCategory(byCategory)
            .totalBudget(totalBudget)
            .averageBudget(averageBudget)
            .completionRate(completionRate)
            .build();
    }
    
    /**
     * Convert Project entity to ProjectDTO
     */
    private ProjectDTO convertToDTO(Project project) {
        return ProjectDTO.builder()
            .id(project.getId())
            .title(project.getTitle())
            .description(project.getDescription())
            .status(project.getStatus())
            .ownerId(project.getOwnerId())
            .budget(project.getBudget())
            .deadline(project.getDeadline())
            .category(project.getCategory())
            .imageUrl(project.getImageUrl())
            .tags(project.getTags())
            .createdAt(project.getCreatedAt())
            .updatedAt(project.getUpdatedAt())
            .build();
    }
}

File: frontend/src/app/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

/**
 * Project Interface
 * Matches the ProjectDTO from backend
 */
export interface Project {
  id?: number;
  title: string;
  description?: string;
  status: 'DRAFT' | 'OPEN' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';
  ownerId: number;
  budget?: number;
  deadline?: string;
  category?: string;
  imageUrl?: string;
  tags?: string;
  createdAt?: string;
  updatedAt?: string;
}

/**
 * Create Project Request Interface
 */
export interface CreateProjectRequest {
  title: string;
  description?: string;
  status?: 'DRAFT' | 'OPEN' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';
  ownerId: number;
  budget?: number;
  deadline?: string;
  category?: string;
  imageUrl?: string;
  tags?: string;
}

/**
 * Update Project Request Interface
 */
export interface UpdateProjectRequest {
  title?: string;
  description?: string;
  status?: 'DRAFT' | 'OPEN' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';
  budget?: number;
  deadline?: string;
  category?: string;
  imageUrl?: string;
  tags?: string;
}

/**
 * Project Service
 * 
 * Provides methods to interact with the Project Microservice backend.
 */
@Injectable({
  providedIn: 'root'
})
export class ProjectService {
  
  private apiUrl = 'http://localhost:8082/api/projects';  // Backend runs on port 8082
  
  constructor(private http: HttpClient) {}
  
  /**
   * Create a new project
   * POST /api/projects
   */
  createProject(request: CreateProjectRequest): Observable<Project> {
    return this.http.post<Project>(this.apiUrl, request);
  }
  
  /**
   * Get project by ID
   * GET /api/projects/{id}
   */
  getProjectById(id: number): Observable<Project> {
    return this.http.get<Project>(`${this.apiUrl}/${id}`);
  }
  
  /**
   * Get all projects
   * GET /api/projects
   */
  getAllProjects(): Observable<Project[]> {
    return this.http.get<Project[]>(this.apiUrl);
  }
  
  /**
   * Get projects by owner ID
   * GET /api/projects/owner/{ownerId}
   */
  getProjectsByOwner(ownerId: number): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/owner/${ownerId}`);
  }
  
  /**
   * Get projects by status
   * GET /api/projects/status/{status}
   */
  getProjectsByStatus(status: string): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/status/${status}`);
  }
  
  /**
   * Get projects by category
   * GET /api/projects/category/{category}
   */
  getProjectsByCategory(category: string): Observable<Project[]> {
    return this.http.get<Project[]>(`${this.apiUrl}/category/${category}`);
  }
  
  /**
   * Update project
   * PUT /api/projects/{id}
   */
  updateProject(id: number, request: UpdateProjectRequest): Observable<Project> {
    return this.http.put<Project>(`${this.apiUrl}/${id}`, request);
  }
  
  /**
   * Delete project
   * DELETE /api/projects/{id}
   */
  deleteProject(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
